raysect/core/math/quaternion.pyx

```diff
@@ -0,0 +1,556 @@
+
+        The returned quaternion is normalised to have norm length 1.0 - a unit quaternion.
+
+        .. code-block:: pycon
+        
+            >>> a = Quaternion(0, 1, 2, 3)
+            >>> a.normalise()
+            Quaternion(0.0, <0.26726, 0.53452, 0.80178>)
+        """
+
+        cdef double n
+
+        # if current length is zero, problem is ill defined
+        n = self.norm()
+        if n == 0.0:
+
+            raise ZeroDivisionError("A zero length quaternion cannot be normalised as the direction "
+                                    "of a zero length quaternion is undefined.")
+
+        # normalise and rescale quaternion
+        n = 1.0 / n
+
+        return self.mul_scalar(n)
+
+    cpdef Quaternion copy(self):
+        """Returns a copy of this quaternion."""
+
+        return new_quaternion(self.s, self.x, self.y, self.z)
+
+    cpdef AffineMatrix3D to_transform(self):
+        """
+        Transforms the vector with the supplied AffineMatrix3D.
+
+        The vector is transformed by pre-multiplying the vector by the affine
+        matrix.
+
+        .. math::
+
+            \\vec{C} = \\textbf{A} \\times \\vec{B}
+
+        This method is substantially faster than using the multiplication
+        operator of AffineMatrix3D when called from cython code.
+
+        :param AffineMatrix3D m: The affine matrix describing the required coordinate transformation.
+        :return: A new instance of this vector that has been transformed with the supplied Affine Matrix.
+        :rtype: Vector3D
+
+        .. code-block:: pycon
+
+            >>> z = Vector3D(0, 0, 1)
+            >>> y = z.transform(rotate_x(90))
+            >>> y
+            Vector3D(0.0, -1.0, 6.123233995736766e-17)
+        """
+
+        cdef:
+            double qs, qx, qy, qz
+            double qs2, qx2, qy2, qz2
+            double m00, m01, m02, m03
+            double m10, m11, m12, m13
+            double m20, m21, m22, m23
+            double m30, m31, m32, m33
+            double temp1, temp2
+            Quaternion unit_q
+
+        unit_q = self.normalise()
+
+        qs = unit_q.s
+        qx = unit_q.x
+        qy = unit_q.y
+        qz = unit_q.z
+
+        qs2 = qs * qs
+        qx2 = qx * qx
+        qy2 = qy * qy
+        qz2 = qz * qz
+
+        m00 = 1 - 2*qy2 - 2*qz2
+        m01 = 2*qx*qy - 2*qz*qs
+        m02 = 2*qx*qz + 2*qy*qs
+
+        m10 = 2*qx*qy + 2*qz*qs
+        m11 = 1 - 2*qx2 - 2*qz2
+        m12 = 2*qy*qz - 2*qx*qs
+
+        m20 = 2*qx*qz - 2*qy*qs
+        m21 = 2*qy*qz + 2*qx*qs
+        m22 = 1 - 2*qx2 - 2*qy2
+
+        m03 = m13 = m23 = 0.0
+        m30 = m31 = m32 = 0.0
+        m33 = 1.0
+
+        return new_affinematrix3d(m00, m01, m02, m03,
+                                  m10, m11, m12, m13,
+                                  m20, m21, m22, m23,
+                                  m30, m31, m32, m33)
+
+    @classmethod
+    def from_axis_angle(cls, axis, angle):
```

Comment:
Missing object types

Reply:
if the arg types are set, you don't need the isinstance tests.

Reply:
Resolved in a4d9d8a